---
title: "Neandertal sQTL"
author: "Arta"
date: "5/26/2020"
output: html_document
---

# Neandertal sQTL Project

I'm going to try to impart as much of my knowledge about the ins/outs of this project onto you as I can. I'll stick to what's the most relevant to you, and avoid bogging you down with stuff you might already know.

## File directory structure

### Base directory

Here are both the folders into the project as well as some loose files, some of which can be discarded and others are essential.

#### Project directories

`envs/` contains the Snakemake environment file which allows for a reproducible set of software versions and packages. It may look incomplete and that's because it is (something I was going to get to after the project once it filled out). This is pretty simple and it involves just listing what you used with the version number, which you would have to do for the GitHub repo anyway. You can delete it but you'll have to make one again anyway if you intend on keeping the Snakemake workflow.

`metadata/` contains files which are used as inputs into tools/scripts. They've been hardcoded into the workflow so I wouldn't recommend touching these.

`results/` is pretty self-explanatory but I will go through each of the individual files with you so you know what they are:

* `AllTissuesNLIso.png` is the graph which shows the number of counts for each sQTL/transcript combination where, on the X-axis, is the number of instances across all individuals where the homozygous human genotype led to more than 0 counts in expression of the transcript, and the Y-axis is the combined number of instances across all individuals where the heterozygous and the homozygous Neandertal genotype led to greater than 0 counts in expression of that same transcript. It sounds confusing but it actually makes a lot of sense if you think about it.

* `loosenedRestrictions*.txt` are from when we decided to allow sQTL/transcript combinations to filter through where the homozygous human allele did not necessarily have to be exactly 0 across all individuals but could be 10 or fewer. `loosenedRestrictionsGenes.txt` has the corresponding gene symbol to the transcript in a column for each row.

* `NL_iso_GOTerms_DAVID.csv`, `NL_Iso_TopGenes_List.txt`, `NL_IsoGeneList.csv` are all lists of genes that I generated not long ago to find functional categories for our Neandertal sQTLs. You're probably good to toss them. 

* `SeparateTissues.png` is similar to `AllTissuesNLIso.png` except broken up by tissue.

* `TopGenes_PermPass_All.csv` are the top sQTLs derived from GTEx's leafcutter results generated by a the script `src/analysis/count_sqtl.R`. More on that later. But these are basically our /results/. 

##### src/

I'm devoting a separate section to `src/` for the simple reason that this is where all of our code lives. Again, many of these scripts are hardcoded into the still-incomplete Snakemake workflow, so I don't recommend moving anything around (especially in `analysis/`). 

`analysis/` is our most important subdirectory since that contains the most active code. `sprime/` comes in second although I believe Rajiv wants to use IBDMix instead so you may end up scrapping that. The other two subdirectories (`splice_site_annotation/` and `sqtl_enrichment/`) I never worked with and could be tossed, as they are dead-ends as far as I know.

Let's go through each script in `analsys/`

* `annotate_gts.py`: This is for the preprocessing of the GTEx v8 VCF (see `preprocessVCF` in the base-level directory for added context)

* `count_sqtl.R` generates a QQ plot and a table of the lowest p-value sQTL. 

* `countCounts.R` counts the number of instances for each variant/sQTL combo where counts are >/< 0 for each tissue. These files may seem out of order in my explanation and that's fine; just look at the DAG at the base directory or the Snakefile.

* `merge_tables.R` brings together the introns counts information from GTEx and sQTL information and merges them together on the basis of transcript ID, variant ID, and subject ID. 

* `NE_sQTL.R` generates "permutation table" files, which made more sense when this project was being developed for GTEx v7 but what's ultimately important is that this file generates the inputs for `count_sqtl.R`

* `plot_countCounts.R` generates `AllTissuesNLIso.png` and `SeparateTissues.png`

* `preprocess_intronCounts.R` takes intron files from GTEx and makes them useable for `merge_tables.R`

* `qqplot.R` makes QQ plots of all of the sQTL across all tissues.

Just a brief overview. It's been a while since I wrote some of these scripts so if there are any errors in my explanation, let me know and I'll take a quick look-see. 

There's not much to explain about `sprime/` other than there is a copy of the whole pipeline in v7 when I was testing things out because the sQTL for ADAMTSL3 disappeared. You can get rid of the v7 files. Again, these scripts are hardcoded into the Snakemake pipeline so if you plan on using them, do not move them around, unless you're not, in which case, feel free to delete.

#### Base directory files

I saved the best for last. I'll try to make this short. 

Scripts
* `geneGOTerms.R` is the script I used to query DAVID for functional categories and then write them to file. The reason they're in the base directory is because I didn't know if it was going to be part of the workflow or not. I don't think it will be so you're free to get rid of it.

* `IBDMix.sh` is the script Rajiv wrote very recently after he discovered he could make IBDMix work. I've done nothing with it other than stick it in the repo.

* `liftover_batch.sh` lifts over the Neandertal genome from hg19 to hg38. In fact, it was part of the IBDMix pipeline that I worked on when we first considered switching from SPrime to IBDMix, but I deleted the IBDMix files except for this one. Interesting. 

* `preprocessVCF`, `*Snakefile`, `config.yaml`, `sf_wrap.sh` are all for Snakemake. `DAG.svg` is a visual representation of the workflow I've been able to build using Snakemake so far. I can't really explain it to you and I didn't really do a great job of annotating the scripts with comments but I can walk you through it if you need me to at some point. 

* `sqtl_notebook.html` is the Rmd file that Steph (quite possibly you) made for me when explaining how were going to get the Neandertal-specific isoforms.

* `Ne_sQTL_Notebook.Rmd` is a new-ish log that I started in January 2020 plotting my progress and regular work.

* `log.md` is an old markdown log I started almost 2 years ago. Damn.